src/extension/app-schema/app-schema-test/src/test/java/org/geoserver/test/PolymorphismWfsTest.java

```diff
@@ -702,4 +719,107 @@ private void checkAnyType(Document doc) {
+
+        // filter must be "unrolled" (i.e. reverse mapped) first
+        Filter unrolled = AppSchemaDataAccess.unrollFilter(equalsFirstValue, rootMapping);
+
+        // Filter is nested
+        assertTrue(NestedFilterToSQL.isNestedFilter(unrolled));
+
+        assertContainsFeatures(fs.getFeatures(equalsFirstValue), "f1");
+
+        /*
+         * test equals filter on second nested attribute
+         */
+        PropertyIsEqualTo equalsSecondValue = ff.equals(
+                ff.property("ex:secondValue/gsml:CGI_NumericValue/gsml:principalValue/@uom"),
+                ff.literal("m"));
+
+        // Filter involving conditional polymorphism --> CANNOT be encoded
+        ComplexFilterSplitter splitterSecondValue = new ComplexFilterSplitter(
+                store.getFilterCapabilities(), rootMapping);
+        splitterSecondValue.visit(equalsSecondValue, null);
+        preFilter = splitterSecondValue.getFilterPre();
+        postFilter = splitterSecondValue.getFilterPost();
+
+        assertEquals(Filter.INCLUDE, preFilter);
+        assertEquals(equalsSecondValue, postFilter);
+
+        // filter must be "unrolled" (i.e. reverse mapped) first
+        unrolled = AppSchemaDataAccess.unrollFilter(equalsSecondValue, rootMapping);
+
+        // Filter is nested
+        assertTrue(NestedFilterToSQL.isNestedFilter(unrolled));
+
+        assertContainsFeatures(fs.getFeatures(equalsSecondValue), "f1", "f3");
+
+        /*
+         * test equals filter on third nested attribute
+         */
+        PropertyIsEqualTo equalsThirdValue = ff.equals(
+                ff.property("ex:thirdValue/gsml:CGI_NumericValue/gsml:principalValue"),
+                ff.literal(1.0));
+
+        // Filter involving non-conditional polymorphism --> can be encoded, because target attribute mapping can be unequivocally determined a-priori
+        ComplexFilterSplitter splitterThirdValueFilter = new ComplexFilterSplitter(
+                store.getFilterCapabilities(), rootMapping);
+        splitterThirdValueFilter.visit(equalsThirdValue, null);
+        preFilter = splitterThirdValueFilter.getFilterPre();
+        postFilter = splitterThirdValueFilter.getFilterPost();
+
+        assertEquals(equalsThirdValue, preFilter);
+        assertEquals(Filter.INCLUDE, postFilter);
+
+        // filter must be "unrolled" (i.e. reverse mapped) first
+        unrolled = AppSchemaDataAccess.unrollFilter(equalsThirdValue, rootMapping);
+
+        // Filter is nested
+        assertTrue(NestedFilterToSQL.isNestedFilter(unrolled));
+
+        NestedFilterToSQL nestedFilterEncoder = createNestedFilterEncoder(rootMapping);
+        String encodedFilter = nestedFilterEncoder.encodeToString(unrolled);
+
+        // this is the generated query in PostGIS, but the test limits to check the presence of the
+        // EXISTS keyword, as the actual SQL is dependent on the underlying database
+//        (EXISTS (SELECT "chain_link_1"."PKEY" 
```

Comment:
Indentation. These comments are good and need to look like code comments so they are preserved and not mistaken for dead code.